### PR TITLE
260 bicgstab smoke tests sometimes does not report useful time 1

### PR DESCRIPTION
**Pull request by anyzelman**
As with issue #166, but now also detected for the BiCGstab smoke test, if the solver fails to converge (and thus returns the `grb::FAILED` error code), the useful time is not recorded.

This MR fixes the issue for BiCGstab, and reviews the smoke tests for all iterative solvers so to make sure the same issue does not persist elsewhere. This MR also makes uniform the implementation of all smoke tests for iterative solvers, and makes the maximum number of iterations a command-line argument for all iterative solver smoke tests. As always, it includes minor code style fixes.

### CODE REVIEW

File: tests/smoke/bicgstab.cpp [hunk 1]
```diff
@@ -28,20 +28,24 @@
...
 
-#define TOL 0.0001
-#define MAX_ITERS 10000
+using namespace grb;
+using namespace algorithms;
 
-#define C1 0.001
-#define C2 0.001
 
+constexpr double TOL = 0.0001;
```
Comment (anyzelman) [inline tests/smoke/bicgstab.cpp]: Lower case variable names

File: tests/smoke/bicgstab.cpp [hunk 2]
```diff
@@ -28,20 +28,24 @@
...
 
-using namespace grb;
-using namespace algorithms;
+/** Default maximum number of solver iterations */
+constexpr size_t MAX_ITERS = 10000;
+
+constexpr double C1 = 0.001;
+
+constexpr double C2 = 0.001;
```
Comment (anyzelman) [inline tests/smoke/bicgstab.cpp]: lower case

File: tests/smoke/bicgstab.cpp [hunk 3]
```diff
@@ -143,17 +147,20 @@ void grbProgram( const struct input &data_in, struct output &out ) {
...
 		);
 		double single_time = timer.time();
-		if( rc != SUCCESS ) {
+		if( !(rc == SUCCESS || rc == FAILED) ) {
 			std::cerr << "Failure: call to bicgstab not succeed ("
 				<< toString( rc ) << ")." << std::endl;
 			out.error_code = 20;
 		}
+		if( rc == FAILED ) {
+			std::cout << "Warning: call to conjugate_gradient did not converge\n";
```
Comment (anyzelman) [inline tests/smoke/bicgstab.cpp]: cg -> bicgstab

File: tests/smoke/conjugate_gradient.cpp [hunk 4]
```diff
@@ -32,27 +32,30 @@
...
+
+
 using BaseScalarType = double;
 #ifdef _CG_COMPLEX
  using ScalarType = std::complex< BaseScalarType >;
 #else
  using ScalarType = BaseScalarType;
 #endif
 
+constexpr BaseScalarType TOL = 0.000001;
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: lower case variable name, also constexpr const

File: tests/smoke/bicgstab.cpp (hunk 1 above)
Comment (anyzelman) [reply tests/smoke/bicgstab.cpp]: also constexpr const

File: tests/smoke/bicgstab.cpp (hunk 2 above)
Comment (anyzelman) [reply tests/smoke/bicgstab.cpp]: also constexpr const

File: tests/smoke/simple_pagerank.cpp [hunk 5]
```diff
@@ -28,14 +28,21 @@
 #include <graphblas.hpp>
 #include <utils/output_verification.hpp>
 
-
 using namespace grb;
 using namespace algorithms;
```
Comment (anyzelman) [inline tests/smoke/simple_pagerank.cpp:33]: extra empty line

File: tests/smoke/simple_pagerank.cpp [hunk 6]
```diff
@@ -28,14 +28,21 @@
...
 
+/** Default maximum number of iterations. */
+constexpr size_t max_iters = 1000;
+
+constexpr double alpha = 0.85;
+
+constexpr double tol = 1e-7;
```
Comment (anyzelman) [inline tests/smoke/simple_pagerank.cpp]: constexpr const

File: tests/smoke/smoketests.sh [hunk 7]
```diff
@@ -205,7 +205,7 @@ for BACKEND in ${BACKENDS[@]}; do
 			echo "                                 variant."
 			echo "Functional test executable: ${TEST_BIN_DIR}/kcore_decomposition_${BACKEND}"
 			if [ -f ${INPUT_DIR}/EPA.mtx ]; then
-				$runner ${TEST_BIN_DIR}/kcore_decomposition_${BACKEND} ${INPUT_DIR}/EPA.mtx direct 1 1 verification ${OUTPUT_VERIFICATION_DIR}/kcore_decomposition_eda_ref &> ${TEST_OUT_DIR}/kcore_decomposition_${BACKEND}_EPA_${P}_${T}.log
+				$runner ${TEST_BIN_DIR}/kcore_decomposition_${BACKEND} ${INPUT_DIR}/EPA.mtx direct 1 1 1000 verification ${OUTPUT_VERIFICATION_DIR}/kcore_decomposition_eda_ref &> ${TEST_OUT_DIR}/kcore_decomposition_${BACKEND}_EPA_${P}_${T}.log
```
Comment (anyzelman) [inline tests/smoke/smoketests.sh]: 1000 should NOT be added here

File: tests/smoke/smoketests.sh [hunk 8]
```diff
@@ -276,7 +276,7 @@ for BACKEND in ${BACKENDS[@]}; do
 			echo "                                 dataset (neurons=1024, layers=120, offset=294) taken from"
 			echo "                                 ${GNN_DATASET_PATH} and using thresholding 32."
 			if [ -d ${GNN_DATASET_PATH} ]; then
-				$runner ${TEST_BIN_DIR}/graphchallenge_nn_single_inference_${BACKEND} ${GNN_DATASET_PATH} 1024 120 294 1 32 indirect 1 1 verification ${OUTPUT_VERIFICATION_DIR}/graphchallenge_nn_out_1024_120_294_32_threshold_ref &> ${TEST_OUT_DIR}/graphchallenge_nn_single_inference_${BACKEND}_${P}_${T}.log
+				$runner ${TEST_BIN_DIR}/graphchallenge_nn_single_inference_${BACKEND} ${GNN_DATASET_PATH} 1024 120 294 1 32 indirect 1 1 1000 verification ${OUTPUT_VERIFICATION_DIR}/graphchallenge_nn_out_1024_120_294_32_threshold_ref &> ${TEST_OUT_DIR}/graphchallenge_nn_single_inference_${BACKEND}_${P}_${T}.log
```
Comment (anyzelman) [inline tests/smoke/smoketests.sh]: 1000 should NOT be added here

File: tests/smoke/smoketests.sh [hunk 9]
```diff
@@ -288,7 +288,7 @@ for BACKEND in ${BACKENDS[@]}; do
 			echo "                                 dataset (neurons=1024, layers=120, offset=294) taken from"
 			echo "                                 ${GNN_DATASET_PATH} and without using thresholding."
 			if [ -d ${GNN_DATASET_PATH} ]; then
-				$runner ${TEST_BIN_DIR}/graphchallenge_nn_single_inference_${BACKEND} ${GNN_DATASET_PATH} 1024 120 294 0 0 indirect 1 1 verification ${OUTPUT_VERIFICATION_DIR}/graphchallenge_nn_out_1024_120_294_no_threshold_ref &> ${TEST_OUT_DIR}/graphchallenge_nn_single_inference_${BACKEND}_${P}_${T}.log
+				$runner ${TEST_BIN_DIR}/graphchallenge_nn_single_inference_${BACKEND} ${GNN_DATASET_PATH} 1024 120 294 0 0 indirect 1 1 1000 verification ${OUTPUT_VERIFICATION_DIR}/graphchallenge_nn_out_1024_120_294_no_threshold_ref &> ${TEST_OUT_DIR}/graphchallenge_nn_single_inference_${BACKEND}_${P}_${T}.log
```
Comment (anyzelman) [inline tests/smoke/smoketests.sh]: 1000 should NOT be added here

File: tests/smoke/smoketests.sh [hunk 10]
```diff
@@ -390,7 +390,7 @@ for BACKEND in ${BACKENDS[@]}; do
 				echo "                                 backend."
 				echo "Functional test executable: ${TEST_BIN_DIR}/kcore_decomposition_critical_${BACKEND}"
 				if [ -f ${INPUT_DIR}/EPA.mtx ]; then
-					$runner ${TEST_BIN_DIR}/kcore_decomposition_critical_${BACKEND} ${INPUT_DIR}/EPA.mtx direct 1 1 verification ${OUTPUT_VERIFICATION_DIR}/kcore_decomposition_eda_ref &> ${TEST_OUT_DIR}/kcore_decomposition_critical_${BACKEND}_EPA_${P}_${T}.log
+					$runner ${TEST_BIN_DIR}/kcore_decomposition_critical_${BACKEND} ${INPUT_DIR}/EPA.mtx direct 1 1 1000 verification ${OUTPUT_VERIFICATION_DIR}/kcore_decomposition_eda_ref &> ${TEST_OUT_DIR}/kcore_decomposition_critical_${BACKEND}_EPA_${P}_${T}.log
```
Comment (anyzelman) [inline tests/smoke/smoketests.sh]: 1000 should NOT be added here

**Comment by anyzelman** [comment]
Review OK. Smoke tests with LPF (MPICH) OK. Waiting CIs

**Comment by anyzelman** [comment]
GitHub CI and internal CI OK, will merge
